iroha/src/sumeragi.rs

```diff
@@ -656,343 +659,561 @@ impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Sumeragi<Q, G, W> {
+            self.sorted_peers.len() > 1
+        }
+
+        /// The minimum number of signatures needed to commit a block
+        pub const fn min_votes_for_commit(&self) -> u32 {
+            2 * self.max_faults + 1
+        }
+
+        /// The minimum number of signatures needed to perform a view change (change leader, proxy, etc.)
+        pub const fn min_votes_for_view_change(&self) -> u32 {
+            self.max_faults + 1
+        }
+
+        /// Peers of set A. They participate in the consensus.
+        pub fn peers_set_a(&self) -> &[PeerId] {
+            let n_a_peers = 2 * self.max_faults + 1;
+            &self.sorted_peers[..n_a_peers as usize]
+        }
+
+        /// Peers of set B. The watch the consensus process.
+        pub fn peers_set_b(&self) -> &[PeerId] {
+            &self.sorted_peers[(2 * self.max_faults + 1) as usize..]
+        }
+
+        /// The leader of the current round.
+        #[allow(clippy::expect_used)]
+        pub fn leader(&self) -> &PeerId {
+            self.peers_set_a()
+                .first()
+                .expect("Failed to get first peer.")
+        }
+
+        /// The proxy tail of the current round.
+        #[allow(clippy::expect_used)]
+        pub fn proxy_tail(&self) -> &PeerId {
+            self.peers_set_a().last().expect("Failed to get last peer.")
+        }
+
+        /// The peers that validate the block in discussion this round and vote for it to be accepted by the blockchain.
+        pub fn validating_peers(&self) -> &[PeerId] {
+            let a_set = self.peers_set_a();
+            if a_set.len() > 1 {
+                &a_set[1..(a_set.len() - 1)]
+            } else {
+                &[]
+            }
+        }
+
+        /// Get role of the peer by its id.
+        pub fn role(&self, peer_id: &PeerId) -> Role {
+            if self.leader() == peer_id {
+                Role::Leader
+            } else if self.proxy_tail() == peer_id {
+                Role::ProxyTail
+            } else if self.validating_peers().contains(peer_id) {
+                Role::ValidatingPeer
+            } else {
+                Role::ObservingPeer
+            }
+        }
+
+        /// Verifies that this `message` was signed by the `signature` of a peer with specified `role`.
+        ///
+        /// # Errors
+        /// Fails if there are no such peer with this key and if signature verification fails
+        pub fn verify_signature_with_role(
+            &self,
+            signature: &Signature,
+            role: Role,
+            message_payload: &[u8],
+        ) -> Result<()> {
+            if role
+                .peers(self)
+                .iter()
+                .any(|peer| peer.public_key == signature.public_key)
+            {
+                Ok(())
+            } else {
+                Err(error!("No {:?} with this public key exists.", role))
+            }
+            .and(signature.verify(message_payload))
```

Comment:
`.and_then` i guess

Reply:
Yep, I think clippy will tell me this :)